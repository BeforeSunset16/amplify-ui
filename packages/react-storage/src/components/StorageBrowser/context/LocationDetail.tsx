import React from 'react';

export interface ActionContext<T extends string> {
  actions: T[];
  selectedAction: T | undefined;
  updateSelectedAction: (value: T | undefined) => void;
}

// think just passed in during "createPaginateContext"
// pageSize: number;
export interface PaginateContext {
  currentPage: number;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  updateCurrentPage: (value: number) => void;
}

const _DetailProvider = ({ children }: { children?: React.ReactNode }) => {
  return <form>{children}</form>;
};
