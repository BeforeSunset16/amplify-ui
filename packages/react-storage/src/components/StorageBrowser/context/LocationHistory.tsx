import React from 'react';
import { ActionState, useDataState } from '@aws-amplify/ui-react-core';

export interface HistoryState {
  currentValue: string;
  previous: string[];
}

export type UpdateHistoryStateActionInput =
  | { currentValue: string; type: 'enter' }
  | { currentValue: never; type: 'exit' };

export type UseHistoryState = [
  ActionState<HistoryState>,
  (input: UpdateHistoryStateActionInput) => void,
];

export const updateHistoryStateAction = (
  prevState: HistoryState,
  { currentValue, type }: UpdateHistoryStateActionInput
): HistoryState => {
  switch (type) {
    case 'enter': {
      const { currentValue: _currentValue, previous } = prevState;
      return { currentValue, previous: [_currentValue, ...previous] };
    }
    case 'exit': {
      const [currentValue, ...previous] = prevState.previous;
      return { currentValue, previous };
    }

    default: {
      // eslint-disable-next-line no-console
      console.error(
        `\`useStorageBrowser('locationHistory')\`: Invalid value of ${type} provided as \`type\``
      );

      return prevState;
    }
  }
};

const HistoryStateContext = React.createContext<UseHistoryState | undefined>(
  undefined
);

export const useHistoryState = (): UseHistoryState => {
  const context = React.useContext(HistoryStateContext);
  if (!context) throw new Error('Must be called inside PROVIDER_NAME_HERE');
  return context;
};

export const HistoryStateProvider = ({
  children,
}: {
  children?: React.ReactNode;
}): JSX.Element => {
  const state = useDataState(updateHistoryStateAction, {
    currentValue: 'Home',
    previous: [],
  });

  return (
    <HistoryStateContext.Provider value={state}>
      {children}
    </HistoryStateContext.Provider>
  );
};
