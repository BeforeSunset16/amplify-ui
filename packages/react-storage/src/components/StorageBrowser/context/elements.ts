import {
  PrimitiveButton as ButtonElementBase,
  PrimitiveView as ViewPrimitiveBase,
  createPrimitivesContext as createElementsContext,
  DefaultPrimitives as ElementsBase,
} from '@aws-amplify/ui-react/internal';

export interface StorageBrowserElements
  extends Pick<ElementsBase, 'Button' | 'View'> {}

const defaultValue: StorageBrowserElements = {
  Button: ButtonElementBase,
  View: ViewPrimitiveBase,
};

export const {
  PrimitivesProvider: ElementsProvider,
  usePrimitive: useElement,
} = createElementsContext(defaultValue);
