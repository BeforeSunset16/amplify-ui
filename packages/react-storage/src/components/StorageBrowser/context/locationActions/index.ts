export { LocationActionsDefault, LOCATION_ACTIONS_DEFAULT } from './defaults';
export {
  LocationActionsContext,
  LocationActionsProvider,
} from './locationActions';
export { LocationAction, LocationActions } from './types';
export { resolveLocationActions } from './utils';
