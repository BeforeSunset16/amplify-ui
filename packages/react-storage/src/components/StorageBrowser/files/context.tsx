import React from 'react';

import { isString, isUndefined } from '@aws-amplify/ui';
import { createContextUtilities } from '@aws-amplify/ui-react-core';
import {
  HandleFileSelect,
  useFileSelect,
} from '@aws-amplify/ui-react/internal';

import { SelectionType } from '../actions/configs';

// @todo add actual error message
const ERROR_MESSAGE = 'Insert meaningful error message here';

type FilesActionType =
  | { type: 'ADD_FILE_ITEMS'; files?: File[] }
  | { type: 'REMOVE_FILE_ITEM'; key: string }
  | { type: 'RESET_FILE_ITEMS' };

type HandleFilesAction = (
  input:
    | FilesActionType
    | { type: 'SELECT_FILES'; selectionType?: SelectionType }
) => void;

export type FileItem = { key: string; item: File };
type FileItems = FileItem[];

export type FilesContextType = [
  files: FileItems,
  handleFileSelect: HandleFilesAction,
];

export const { FilesContext, useFiles } = createContextUtilities<
  FilesContextType,
  'Files'
>({ contextName: 'Files', errorMessage: ERROR_MESSAGE });

const exclude = (items: FileItems, target: FileItem) =>
  items.some(({ key }) => key === target.key);

const resolveFiles = (
  prevItems: FileItems,
  files: File[] | undefined
): FileItems => {
  if (!files?.length) return prevItems;

  const incomingItems = files.map((file) => ({ key: file.name, item: file }));

  if (!prevItems.length) return incomingItems;

  return [...prevItems, ...incomingItems].reduce(
    (output, next) => (exclude(output, next) ? output : [...output, next]),
    [] as FileItems
  );
};

const filesReducer: React.Reducer<FileItems, FilesActionType> = (
  prevItems,
  input
) => {
  switch (input.type) {
    case 'ADD_FILE_ITEMS': {
      return resolveFiles(prevItems, input.files);
    }
    case 'REMOVE_FILE_ITEM': {
      return prevItems.filter(({ key }) => key !== input.key);
    }
    case 'RESET_FILE_ITEMS': {
      return [];
    }
  }
};

const parseFileSelectParams = (
  value?: SelectionType
): Parameters<HandleFileSelect> => {
  if (isUndefined(value)) return ['FILE', undefined];
  if (isString(value)) return [value, undefined];

  const [selectType, ...rest] = value;
  return [selectType, isUndefined(rest) ? rest : { accept: rest.join() }];
};

export function FilesProvider({
  children,
}: {
  children?: React.ReactNode;
}): React.JSX.Element {
  const [items, dispatch] = React.useReducer(filesReducer, []);

  const [fileInput, handleFileSelect] = useFileSelect((nextFiles) => {
    dispatch({ type: 'ADD_FILE_ITEMS', files: nextFiles });
  });

  const handleFilesAction: HandleFilesAction = React.useCallback(
    (action) => {
      if (action.type === 'SELECT_FILES') {
        handleFileSelect(...parseFileSelectParams(action.selectionType));
        return;
      }
      return dispatch(action);
    },
    [handleFileSelect]
  );

  const value: FilesContextType = React.useMemo(
    () => [items, handleFilesAction],
    [items, handleFilesAction]
  );

  return (
    <FilesContext.Provider value={value}>
      {fileInput}
      {children}
    </FilesContext.Provider>
  );
}
