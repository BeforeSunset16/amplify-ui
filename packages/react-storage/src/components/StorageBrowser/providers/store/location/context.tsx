import React from 'react';

import { createContextUtilities } from '@aws-amplify/ui-react-core';
import { noop } from '@aws-amplify/ui';

import { assertIsLocationData } from '../../../validators';
import { LocationData } from '../../../actions';

export const ERROR_MESSAGE =
  'Invalid `location` value provided as initial value to `LocationProvider.';

export const DEFAULT_STATE: LocationState = {
  location: undefined,
  path: undefined,
};

export type LocationActionType =
  | { type: 'NAVIGATE'; location: LocationData; path?: string }
  | { type: 'RESET_LOCATION' };

/**
 * Stores current `location` and nested `path` values set on
 * navigate events.
 *
 * @example
 *`"s3//:my-bucket/my-prefix/a-path/a-nested-path/"`
 *
 * ```
 * const state: LocationState = {
 *   location: {
 *     bucket: 'my-bucket',
 *     permisson: 'READWRITE',
 *     prefix: 'my-prefix/',
 *     type: 'PREFIX',
 *   },
 *   path: 'a-path/a-nested-path/',
 * }
 *
 */
export interface LocationState {
  /**
   * current `location` metadata
   */
  location: LocationData | undefined;

  /**
   * current `location` subpath within a `location`
   */
  path: string | undefined;
}

export type LocationStateContext = [
  LocationState,
  (action: LocationActionType) => void,
];

export interface LocationProviderProps extends Partial<LocationState> {
  children?: React.ReactNode;
}

function handleAction(
  state: LocationState,
  action: LocationActionType
): LocationState {
  switch (action.type) {
    case 'NAVIGATE': {
      const { location, path } = action;

      if (state.location?.id === location.id && state.path === path) {
        return state;
      }

      return { location, path };
    }
    case 'RESET_LOCATION': {
      return DEFAULT_STATE;
    }
  }
}

const defaultValue: LocationStateContext = [DEFAULT_STATE, noop];
export const { LocationContext, useLocation } = createContextUtilities({
  contextName: 'Location',
  defaultValue,
});

export function LocationProvider({
  children,
  location,
  path,
}: LocationProviderProps): React.JSX.Element {
  if (location) {
    assertIsLocationData(location, ERROR_MESSAGE);
  }

  const value = React.useReducer(
    handleAction,
    location ? { location, path } : DEFAULT_STATE
  );

  return (
    <LocationContext.Provider value={value}>
      {children}
    </LocationContext.Provider>
  );
}
