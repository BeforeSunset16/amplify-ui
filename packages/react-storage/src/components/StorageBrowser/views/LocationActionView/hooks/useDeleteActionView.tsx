import { humanFileSize, isFunction } from '@aws-amplify/ui';

import { deleteHandler } from '../../../actions/handlers';
import { TaskCounts } from '../../../controls/types';
import {
  DataTableProps,
  DataTableRow,
} from '../../../composables/DataTable/DataTable';
import { WithKey } from '../../../components/types';
import { getTaskCounts } from '../../../controls/getTaskCounts';
import { STATUS_DISPLAY_VALUES } from '../../LocationActionView/constants';
import { useStore } from '../../../providers/store';
import { useGetActionInput } from '../../../providers/configuration';
import { Task, useProcessTasks } from '../../../tasks';
import { UseActionView } from './types';
import {
  getActionIconVariant,
  getActionViewTaskStatuses,
  getFilenameWithoutPrefix,
  getFileTypeDisplayValue,
} from './utils';

interface UseDeleteActionView extends UseActionView {}

const LOCATION_ACTION_VIEW_HEADERS: DataTableProps['headers'] = [
  { key: 'key', type: 'sort', content: { label: 'Name' } },
  { key: 'folder', type: 'text', content: { text: 'Folder' } },
  { key: 'type', type: 'text', content: { text: 'Type' } },
  { key: 'size', type: 'text', content: { text: 'Size' } },
  { key: 'status', type: 'sort', content: { label: 'Status' } },
  { key: 'action', type: 'text', content: { text: '' } },
];

export const getDeleteActionViewTableData = ({
  tasks,
  taskCounts,
  path,
}: {
  tasks: Task[];
  taskCounts: TaskCounts;
  path: string;
}): DataTableProps => {
  const { hasStarted } = getActionViewTaskStatuses(taskCounts);

  const rows: DataTableProps['rows'] = tasks.map((item) => {
    const row: WithKey<DataTableRow> = {
      key: item.key,
      content: LOCATION_ACTION_VIEW_HEADERS.map(({ key: columnKey }) => {
        const key = `${columnKey}-${item.key}`;
        switch (columnKey) {
          case 'key': {
            return {
              key,
              type: 'text',
              content: {
                icon: getActionIconVariant(item.status),
                text: getFilenameWithoutPrefix(item.key),
              },
            };
          }
          case 'folder': {
            return {
              key,
              type: 'text',
              content: {
                text: path,
              },
            };
          }
          case 'type': {
            return {
              key,
              type: 'text',
              content: {
                text: getFileTypeDisplayValue(item.key),
              },
            };
          }
          case 'size':
            return {
              key,
              type: 'text',
              content: {
                text: humanFileSize(
                  parseInt((item as Task & { size: string }).size),
                  true
                ),
              },
            };
          case 'status':
            return {
              key,
              type: 'text',
              content: {
                text: STATUS_DISPLAY_VALUES[item.status],
              },
            };
          case 'action':
            // don't allow removing a single task
            if (taskCounts.TOTAL > 1) {
              return hasStarted
                ? {
                    key,
                    type: 'button',
                    content: {
                      icon: 'cancel',
                      ariaLabel: `Cancel item: ${item.key}`,
                      onClick: () => item.cancel?.(),
                      disabled: item.status !== 'QUEUED',
                    },
                  }
                : {
                    key,
                    type: 'button',
                    content: {
                      icon: 'cancel',
                      ariaLabel: `Remove item: ${item.key}`,
                      onClick: () => item.remove(),
                    },
                  };
            } else {
              return {
                key,
                type: 'text',
                content: {
                  text: '',
                },
              };
            }
          default:
            return {
              key,
              type: 'text',
              content: {
                text: '',
              },
            };
        }
      }),
    };
    return row;
  });

  const tableData: DataTableProps = {
    headers: LOCATION_ACTION_VIEW_HEADERS,
    rows,
  };
  return tableData;
};

export const useDeleteActionView = ({
  onClose: _onClose,
}: {
  onClose?: () => void;
}): UseDeleteActionView => {
  const [
    {
      history,
      locationItems: { fileDataItems: selected },
    },
    dispatchStoreAction,
  ] = useStore();
  const { current } = history;
  const path = current?.prefix;
  const getInput = useGetActionInput();

  const [tasks, handleProcess] = useProcessTasks(
    // @ts-expect-error
    deleteHandler,
    selected,
    {
      concurrency: 1,
    }
  );

  const taskCounts = getTaskCounts(tasks);
  const { disableCancel, disableClose, disablePrimary } =
    getActionViewTaskStatuses(taskCounts);

  const tableData = getDeleteActionViewTableData({
    tasks,
    taskCounts,
    path: path ?? '',
  });

  const onStart = () => {
    if (!current?.prefix) return;
    handleProcess({
      config: getInput(),
      prefix: current.prefix,
    });
  };

  const onCancel = () => {
    tasks.forEach((task) => {
      // @TODO Fixme, calling cancel on task doesn't currently work
      if (isFunction(task.cancel)) task.cancel();
    });
  };

  const onClose = () => {
    // clear files state
    dispatchStoreAction({ type: 'RESET_LOCATION_ITEMS' });
    // clear selected action
    dispatchStoreAction({ type: 'RESET_ACTION_TYPE' });
    if (isFunction(_onClose)) _onClose();
  };

  return {
    disableCancel,
    disableClose,
    disablePrimary,
    onCancel,
    onClose,
    onStart,
    taskCounts,
    tableData,
  };
};
