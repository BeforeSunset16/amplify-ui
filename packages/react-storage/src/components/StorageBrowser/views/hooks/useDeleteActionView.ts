import { useMemo, useState } from 'react';

import { isFunction } from '@aws-amplify/ui';

import { useControl } from '../../context/control';
import { deleteHandler } from '../../actions/handlers';
import { Task, useProcessTasks } from '../../tasks';
import { useGetLocationConfig } from '../../context/config';
import { UseActionView } from './types';

interface UseDeleteActionView extends UseActionView {
  path: string;
  tasks: Task[];
}

export const UseDeleteActionView = (): UseDeleteActionView => {
  const [isProcessing, setIsProcessing] = useState(false);
  const getConfig = useGetLocationConfig();
  const { bucket, credentialsProvider, region } = getConfig();

  const [{ path }] = useControl('NAVIGATE');
  const [, handleUpdateState] = useControl('LOCATION_ACTIONS');
  const [
    {
      selected: { items: selected = [] },
    },
  ] = useControl('LOCATION_ACTIONS');

  const processTasksInputItems = useMemo(() => {
    return selected
      ? selected.map((item) => ({
          ...item,
          key: item.key,
          item: {},
          cancel: () => {},
        }))
      : [];
  }, [selected]);

  const [tasks, processTasks] = useProcessTasks(
    deleteHandler,
    processTasksInputItems
  );

  const onStart = () => {
    processTasks({
      prefix: path ?? '',
      config: {
        accountId: '', // @TODO: pass in accountId here
        bucket,
        credentials: credentialsProvider,
        region,
      },
    });
    setIsProcessing(true);
  };

  const onCancel = () => {
    tasks.forEach((task) => {
      // @TODO Fixme, cancel doesn't currently cancel the task
      if (isFunction(task.cancel)) task.cancel();
    });
  };

  const onClose = () => {
    handleUpdateState({ type: 'CLEAR' });
  };

  return {
    path: path ?? '',
    tasks,
    isProcessing,
    onCancel,
    onClose,
    onStart,
  };
};
