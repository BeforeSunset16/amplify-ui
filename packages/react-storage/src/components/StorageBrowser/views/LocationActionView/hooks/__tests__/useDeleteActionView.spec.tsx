import { renderHook, act } from '@testing-library/react-hooks';
import { UseDeleteActionView } from '../useDeleteActionView';
import { useGetLocationConfig } from '../../../../context/config';
import { useControl } from '../../../../context/control';
import { useProcessTasks } from '../../../../tasks';
import { getDeleteActionViewTableData } from '../useDeleteActionView';
import { TaskStatus } from '../../../../tasks';

// Mock the imported hooks and functions
jest.mock('../../../../context/config');
jest.mock('../../../../context/control');
jest.mock('../../../../tasks');
jest.mock('../../../../actions/handlers');

describe('UseDeleteActionView', () => {
  beforeEach(() => {
    // Mock the useGetLocationConfig hook
    (useGetLocationConfig as jest.Mock).mockReturnValue(() => ({
      bucket: 'XXXXXXXXXXX',
      credentialsProvider: 'test-credentials',
      region: 'us-west-2',
      accountId: '123456789012',
    }));

    // Mock the useControl hook
    (useControl as jest.Mock).mockReturnValue([
      { path: 'test-path', selected: { items: [{ key: 'test-item' }] } },
      jest.fn(),
    ]);

    // Mock the useProcessTasks hook
    (useProcessTasks as jest.Mock).mockReturnValue([
      [
        { key: 'test-item', status: 'QUEUED' },
        { key: 'test-item2', status: 'QUEUED' },
        { key: 'test-item3', status: 'QUEUED' },
      ],
      jest.fn(),
    ]);
  });

  it('should return the correct initial state', () => {
    const { result } = renderHook(() => UseDeleteActionView());

    expect(result.current).toEqual(
      expect.objectContaining({
        disableCancel: true,
        disableClose: false,
        disablePrimary: false,
        onCancel: expect.any(Function),
        onClose: expect.any(Function),
        onStart: expect.any(Function),
      })
    );

    expect(result.current.controlsContextValue).toEqual(
      expect.objectContaining({
        data: expect.any(Object),
        actionsConfig: { type: 'BATCH_ACTION', isCancelable: true },
      })
    );
  });

  it('should call processTasks when onStart is called', () => {
    const mockProcessTasks = jest.fn();
    (useProcessTasks as jest.Mock).mockReturnValue([[], mockProcessTasks]);

    const { result } = renderHook(() => UseDeleteActionView());

    act(() => {
      result.current.onStart();
    });

    expect(mockProcessTasks).toHaveBeenCalledWith({
      prefix: 'test-path',
      config: {
        accountId: '123456789012',
        bucket: 'XXXXXXXXXXX',
        credentials: 'test-credentials',
        region: 'us-west-2',
      },
    });
  });

  it('should call cancel on tasks when onCancel is called', () => {
    const mockCancel = jest.fn();
    (useProcessTasks as jest.Mock).mockReturnValue([
      [{ key: 'test-item', status: 'PENDING', cancel: mockCancel }],
      jest.fn(),
    ]);

    const { result } = renderHook(() => UseDeleteActionView());

    act(() => {
      result.current.onCancel();
    });

    expect(mockCancel).toHaveBeenCalled();
  });

  it('should clear the state when onClose is called', () => {
    const mockUpdateState = jest.fn();
    (useControl as jest.Mock).mockReturnValue([
      { path: 'test-path', selected: { items: [{ key: 'test-item' }] } },
      mockUpdateState,
    ]);

    const { result } = renderHook(() => UseDeleteActionView());

    act(() => {
      result.current.onClose();
    });

    expect(mockUpdateState).toHaveBeenCalledWith({ type: 'CLEAR' });
  });

  it('should disable close and primary when some tasks in progress', () => {
    (useProcessTasks as jest.Mock).mockReturnValue([
      [
        { key: 'item1', status: 'QUEUED' },
        { key: 'item1', status: 'COMPLETE' },
        { key: 'item2', status: 'PENDING' },
      ],
      jest.fn(),
    ]);

    const { result } = renderHook(() => UseDeleteActionView());

    expect(result.current).toEqual(
      expect.objectContaining({
        disableCancel: false,
        disableClose: true,
        disablePrimary: true,
      })
    );
  });

  it('should disable cancel, close and primary when all tasks in progress or complete', () => {
    (useProcessTasks as jest.Mock).mockReturnValue([
      [
        { key: 'item1', status: 'PENDING' },
        { key: 'item2', status: 'PENDING' },
        { key: 'item1', status: 'COMPLETE' },
      ],
      jest.fn(),
    ]);

    const { result } = renderHook(() => UseDeleteActionView());

    expect(result.current).toEqual(
      expect.objectContaining({
        disableCancel: true,
        disableClose: true,
        disablePrimary: true,
      })
    );
  });
});

describe('getDeleteActionViewTableData', () => {
  const mockRemove = jest.fn();
  const taskCounts = {
    INITIAL: 0,
    QUEUED: 1,
    PENDING: 1,
    FAILED: 1,
    COMPLETE: 1,
    CANCELED: 1,
    TOTAL: 5,
  };
  const tasks = [
    {
      key: 'file1.txt',
      status: 'QUEUED' as TaskStatus,
      remove: mockRemove,
      cancel: jest.fn(),
      item: {},
      message: '',
      size: 1000,
    },
    {
      key: 'file2.jpg',
      status: 'PENDING' as TaskStatus,
      remove: mockRemove,
      cancel: jest.fn(),
      item: {},
      message: '',
      size: 1000,
    },
    {
      key: 'file3.pdf',
      status: 'COMPLETE' as TaskStatus,
      remove: mockRemove,
      cancel: jest.fn(),
      item: {},
      message: '',
      size: 1000,
    },
    {
      key: 'file4.doc',
      status: 'FAILED' as TaskStatus,
      remove: mockRemove,
      cancel: jest.fn(),
      item: {},
      message: '',
      size: 1000,
    },
    {
      key: 'file5',
      status: 'CANCELED' as TaskStatus,
      remove: mockRemove,
      cancel: jest.fn(),
      item: {},
      message: '',
      size: 1000,
    },
  ];

  it('should return correct table data for all task statuses', () => {
    const result = getDeleteActionViewTableData({
      tasks,
      taskCounts,
      path: '',
    });

    expect(result.rows).toMatchSnapshot('tabledata');
  });

  it('should handle tasks with prefix keys', () => {
    const taskCounts = {
      INITIAL: 0,
      QUEUED: 1,
      PENDING: 0,
      FAILED: 0,
      COMPLETE: 1,
      CANCELED: 0,
      TOTAL: 2,
    };
    const tasksWithPaths = [
      {
        key: 'folder/subfolder/file1.txt',
        status: 'QUEUED' as TaskStatus,
        remove: mockRemove,
        cancel: jest.fn(),
        item: {},
        message: '',
        size: 1000,
      },
      {
        key: '/root/file2.jpg',
        status: 'COMPLETE' as TaskStatus,
        remove: mockRemove,
        cancel: jest.fn(),
        item: {},
        message: '',
        size: 1000,
      },
    ];

    const result = getDeleteActionViewTableData({
      tasks: tasksWithPaths,
      taskCounts,
      path: '',
    });

    expect(result.rows).toMatchSnapshot();
  });

  it('should have remove handler on queued files', () => {
    const taskCounts = {
      INITIAL: 0,
      QUEUED: 1,
      PENDING: 0,
      FAILED: 0,
      COMPLETE: 1,
      CANCELED: 0,
      TOTAL: 2,
    };
    const mockRemove = jest.fn();
    const mockCancel = jest.fn();
    const tasksWithPaths = [
      {
        key: 'folder/subfolder/file1.txt',
        status: 'QUEUED' as TaskStatus,
        remove: mockRemove,
        cancel: mockCancel,
        item: {},
        message: '',
        size: 1000,
      },
    ];

    const result = getDeleteActionViewTableData({
      tasks: tasksWithPaths,
      taskCounts,
      path: 'folder/subfolder/',
    });
    const actionCell = result.rows[0].content.filter(
      (cell) => cell.key === 'action-folder/subfolder/file1.txt'
    )[0];
    expect(actionCell.content).toHaveProperty('onClick');
    expect(actionCell.content).toHaveProperty(
      'ariaLabel',
      'Cancel item: folder/subfolder/file1.txt'
    );
    expect(result.rows).toMatchSnapshot();
  });
});
