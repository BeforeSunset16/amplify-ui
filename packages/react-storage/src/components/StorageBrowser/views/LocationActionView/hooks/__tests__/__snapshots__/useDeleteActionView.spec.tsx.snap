// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDeleteActionViewTableData should handle tasks with prefix keys 1`] = `
[
  {
    "content": [
      {
        "content": {
          "icon": "action-queued",
          "text": "file1.txt",
        },
        "key": "key-folder/subfolder/file1.txt",
        "type": "text",
      },
      {
        "content": {
          "text": "",
        },
        "key": "folder-folder/subfolder/file1.txt",
        "type": "text",
      },
      {
        "content": {
          "text": "txt",
        },
        "key": "type-folder/subfolder/file1.txt",
        "type": "text",
      },
      {
        "content": {
          "text": "1.0 kB",
        },
        "key": "size-folder/subfolder/file1.txt",
        "type": "text",
      },
      {
        "content": {
          "text": "Queued",
        },
        "key": "status-folder/subfolder/file1.txt",
        "type": "text",
      },
      {
        "content": {
          "ariaLabel": "Cancel item: folder/subfolder/file1.txt",
          "disabled": false,
          "icon": "cancel",
          "onClick": [Function],
        },
        "key": "action-folder/subfolder/file1.txt",
        "type": "button",
      },
    ],
    "key": "folder/subfolder/file1.txt",
  },
  {
    "content": [
      {
        "content": {
          "icon": "action-success",
          "text": "file2.jpg",
        },
        "key": "key-/root/file2.jpg",
        "type": "text",
      },
      {
        "content": {
          "text": "",
        },
        "key": "folder-/root/file2.jpg",
        "type": "text",
      },
      {
        "content": {
          "text": "jpg",
        },
        "key": "type-/root/file2.jpg",
        "type": "text",
      },
      {
        "content": {
          "text": "1.0 kB",
        },
        "key": "size-/root/file2.jpg",
        "type": "text",
      },
      {
        "content": {
          "text": "Completed",
        },
        "key": "status-/root/file2.jpg",
        "type": "text",
      },
      {
        "content": {
          "ariaLabel": "Cancel item: /root/file2.jpg",
          "disabled": true,
          "icon": "cancel",
          "onClick": [Function],
        },
        "key": "action-/root/file2.jpg",
        "type": "button",
      },
    ],
    "key": "/root/file2.jpg",
  },
]
`;

exports[`getDeleteActionViewTableData should have remove handler on queued files 1`] = `
[
  {
    "content": [
      {
        "content": {
          "icon": "action-queued",
          "text": "file1.txt",
        },
        "key": "key-folder/subfolder/file1.txt",
        "type": "text",
      },
      {
        "content": {
          "text": "folder/subfolder/",
        },
        "key": "folder-folder/subfolder/file1.txt",
        "type": "text",
      },
      {
        "content": {
          "text": "txt",
        },
        "key": "type-folder/subfolder/file1.txt",
        "type": "text",
      },
      {
        "content": {
          "text": "1.0 kB",
        },
        "key": "size-folder/subfolder/file1.txt",
        "type": "text",
      },
      {
        "content": {
          "text": "Queued",
        },
        "key": "status-folder/subfolder/file1.txt",
        "type": "text",
      },
      {
        "content": {
          "ariaLabel": "Cancel item: folder/subfolder/file1.txt",
          "disabled": false,
          "icon": "cancel",
          "onClick": [Function],
        },
        "key": "action-folder/subfolder/file1.txt",
        "type": "button",
      },
    ],
    "key": "folder/subfolder/file1.txt",
  },
]
`;

exports[`getDeleteActionViewTableData should return correct table data for all task statuses: tabledata 1`] = `
[
  {
    "content": [
      {
        "content": {
          "icon": "action-queued",
          "text": "file1.txt",
        },
        "key": "key-file1.txt",
        "type": "text",
      },
      {
        "content": {
          "text": "",
        },
        "key": "folder-file1.txt",
        "type": "text",
      },
      {
        "content": {
          "text": "txt",
        },
        "key": "type-file1.txt",
        "type": "text",
      },
      {
        "content": {
          "text": "1.0 kB",
        },
        "key": "size-file1.txt",
        "type": "text",
      },
      {
        "content": {
          "text": "Queued",
        },
        "key": "status-file1.txt",
        "type": "text",
      },
      {
        "content": {
          "ariaLabel": "Cancel item: file1.txt",
          "disabled": false,
          "icon": "cancel",
          "onClick": [Function],
        },
        "key": "action-file1.txt",
        "type": "button",
      },
    ],
    "key": "file1.txt",
  },
  {
    "content": [
      {
        "content": {
          "icon": "action-progress",
          "text": "file2.jpg",
        },
        "key": "key-file2.jpg",
        "type": "text",
      },
      {
        "content": {
          "text": "",
        },
        "key": "folder-file2.jpg",
        "type": "text",
      },
      {
        "content": {
          "text": "jpg",
        },
        "key": "type-file2.jpg",
        "type": "text",
      },
      {
        "content": {
          "text": "1.0 kB",
        },
        "key": "size-file2.jpg",
        "type": "text",
      },
      {
        "content": {
          "text": "In Progress",
        },
        "key": "status-file2.jpg",
        "type": "text",
      },
      {
        "content": {
          "ariaLabel": "Cancel item: file2.jpg",
          "disabled": true,
          "icon": "cancel",
          "onClick": [Function],
        },
        "key": "action-file2.jpg",
        "type": "button",
      },
    ],
    "key": "file2.jpg",
  },
  {
    "content": [
      {
        "content": {
          "icon": "action-success",
          "text": "file3.pdf",
        },
        "key": "key-file3.pdf",
        "type": "text",
      },
      {
        "content": {
          "text": "",
        },
        "key": "folder-file3.pdf",
        "type": "text",
      },
      {
        "content": {
          "text": "pdf",
        },
        "key": "type-file3.pdf",
        "type": "text",
      },
      {
        "content": {
          "text": "1.0 kB",
        },
        "key": "size-file3.pdf",
        "type": "text",
      },
      {
        "content": {
          "text": "Completed",
        },
        "key": "status-file3.pdf",
        "type": "text",
      },
      {
        "content": {
          "ariaLabel": "Cancel item: file3.pdf",
          "disabled": true,
          "icon": "cancel",
          "onClick": [Function],
        },
        "key": "action-file3.pdf",
        "type": "button",
      },
    ],
    "key": "file3.pdf",
  },
  {
    "content": [
      {
        "content": {
          "icon": "action-error",
          "text": "file4.doc",
        },
        "key": "key-file4.doc",
        "type": "text",
      },
      {
        "content": {
          "text": "",
        },
        "key": "folder-file4.doc",
        "type": "text",
      },
      {
        "content": {
          "text": "doc",
        },
        "key": "type-file4.doc",
        "type": "text",
      },
      {
        "content": {
          "text": "1.0 kB",
        },
        "key": "size-file4.doc",
        "type": "text",
      },
      {
        "content": {
          "text": "Failed",
        },
        "key": "status-file4.doc",
        "type": "text",
      },
      {
        "content": {
          "ariaLabel": "Cancel item: file4.doc",
          "disabled": true,
          "icon": "cancel",
          "onClick": [Function],
        },
        "key": "action-file4.doc",
        "type": "button",
      },
    ],
    "key": "file4.doc",
  },
  {
    "content": [
      {
        "content": {
          "icon": "action-canceled",
          "text": "file5",
        },
        "key": "key-file5",
        "type": "text",
      },
      {
        "content": {
          "text": "",
        },
        "key": "folder-file5",
        "type": "text",
      },
      {
        "content": {
          "text": "",
        },
        "key": "type-file5",
        "type": "text",
      },
      {
        "content": {
          "text": "1.0 kB",
        },
        "key": "size-file5",
        "type": "text",
      },
      {
        "content": {
          "text": "Canceled",
        },
        "key": "status-file5",
        "type": "text",
      },
      {
        "content": {
          "ariaLabel": "Cancel item: file5",
          "disabled": true,
          "icon": "cancel",
          "onClick": [Function],
        },
        "key": "action-file5",
        "type": "button",
      },
    ],
    "key": "file5",
  },
]
`;
